src/assets/scripts/client/aircraft/FlightManagementSystem/RouteModel.js

```diff
@@ -0,0 +1,902 @@
+ * to modify, including adding/removing legs/waypoints, adding/removing waypoint
+ * restrictions, absorbing another `RouteModel`, etc.
+ *
+ * @class RouteModel
+ */
+export default class RouteModel extends BaseModel {
+    /**
+     * @for RouteModel
+     * @constructor
+     * @param navigationLibrary {NavigationLibrary}
+     * @param routeString {string}
+     */
+    constructor(navigationLibrary, routeString) {
+        if (!(navigationLibrary instanceof NavigationLibrary)) {
+            throw new TypeError(`Expected valid navigationLibrary, but received ${typeof navigationLibrary}`);
+        }
+
+        super();
+
+        /**
+         * Array of `LegModel`s on the route
+         *
+         * @for RouteModel
+         * @property _legCollection
+         * @type {array<LegModel>}
+         * @private
+         */
+        this._legCollection = [];
+
+        /**
+         * Local reference to NavigationLibrary
+         *
+         * @for RouteModel
+         * @property _navigationLibrary
+         * @type {NavigationLibrary}
+         * @private
+         */
+        this._navigationLibrary = navigationLibrary;
+
+        // FIXME: Use this
+        /**
+         * Array of `LegModel`s that have been passed (or skipped)
+         *
+         * Aircraft will proceed along the route to each waypoint, and upon completion
+         * of any given leg, it will move that leg here to the `#_previousLegCollection`,
+         * and proceed to the next leg in the `#_legCollection` until no more `LegModel`s
+         * exist, at which point they will simply hold their last assigned heading and altitude.
+         *
+         * @for RouteModel
+         * @property _previousLegCollection
+         * @type {array<WaypointModel>}
+         * @private
+         */
+        this._previousLegCollection = [];
+
+        this.init(routeString);
+    }
+
+    /**
+     * Return the current `LegModel`
+     *
+     * @for RouteModel
+     * @property currentLeg
+     * @type {LegModel}
+     */
+    get currentLeg() {
+        if (this._legCollection.length < 1) {
+            throw new TypeError('Expected the route to contain at least one leg');
+        }
+
+        return this._legCollection[0];
+    }
+
+    /**
+     * Return the current `WaypointModel`
+     *
+     * @for RouteModel
+     * @property currentWaypoint
+     * @type {WaypointModel}
+     */
+    get currentWaypoint() {
+        return this.currentLeg.currentWaypoint;
+    }
+
+    /**
+     * Return the next `LegModel`, if it exists
+     *
+     * @for RouteModel
+     * @property nextLeg
+     * @type {LegModel}
+     */
+    get nextLeg() {
+        if (!this.hasNextLeg()) {
+            return null;
+        }
+
+        return this._legCollection[1];
+    }
+
+    /**
+     * Return the next `WaypointModel`, from current or future leg
+     *
+     * @for RouteModel
+     * @property nextWaypoint
+     * @type {WaypointModel}
+     */
+    get nextWaypoint() {
```

Comment:
there's a lot of logic here for a getter.  this is bordering on method territory.

Reply:
Let’s go with a method